addons/base_address_city/models/res_partner.py

```diff
@@ -5,39 +5,35 @@
 
 from odoo import api, models, fields
 
-class FormatAddressMixin(models.AbstractModel):
-    _inherit = "format.address.mixin"
+class Partner(models.Model):
+    _inherit = 'res.partner'
+
+    country_enforce_cities = fields.Boolean(related='country_id.enforce_cities')
+    city_id = fields.Many2one('res.city', string='Company')
+
+    @api.onchange('city_id')
+    def _onchange_city_id(self):
+        self.city = self.city_id.name
+        self.zip = self.city_id.zipcode
+        self.state_id = self.city_id.state_id
 
     @api.model
-    def fields_view_get_address(self, arch):
-        arch = super(FormatAddressMixin, self).fields_view_get_address(arch)
-        #render the partner address accordingly to address_view_id
+    def _fields_view_get_address(self, arch):
+        arch = super(Partner, self)._fields_view_get_address(arch)
+        if not self._context.get('no_address_format'):
+            return arch
+        # render the partner address accordingly to address_view_id
         doc = etree.fromstring(arch)
         for city_node in doc.xpath("//field[@name='city']"):
             replacement_xml = """
             <div>
-            <field name="country_enforce_cities" invisible="1"/>
-            <div attrs="{'invisible': [('country_enforce_cities', '=', False)]}">
-                <field name='city' attrs="{'invisible': ['|', ('city_id', '!=', False), ('city', '=', False)]}"/>
-                <field name='city_id'/>
-            </div>
+                <field name="country_enforce_cities" invisible="1"/>
+                <field name='city' attrs="{'invisible': [('country_enforce_cities', '=', True), ('city_id', '!=', False)], 'readonly': [('type', '=', 'contact'), ('parent_id', '!=', False)]}"/>
```

Comment:
@moylop260 

What do you think about this logic?

`<field name='city' attrs="{'invisible': [('country_enforce_cities', '=', True), ('city_id', '!=', False)],...`

If city_id is empty, show both fields